### PR TITLE
[mlir][EmitC] Add an `emitc.conditional` operator

### PR DESCRIPTION
**Pull request by marbre**
This adds an `emitc.conditional` operation for the ternary conditional operator. Furthermore, this adds a converion from `arith.select` to the new op.

### CODE REVIEW

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 1]
```diff
@@ -54,6 +54,26 @@ class ArithOpConversion final : public OpConversionPattern<ArithOp> {
...
+
+class SelectOpConversion : public OpConversionPattern<arith::SelectOp> {
+public:
+  using OpConversionPattern<arith::SelectOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(arith::SelectOp selectOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    if (!selectOp.getCondition().getType().isInteger())
```
[the remark this answers is not in the document]
Comment (marbre) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp]: Thanks for fetching this.

**Review by mgehre-amd — APPROVED** [review]

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 2]
```diff
@@ -54,6 +54,31 @@ class ArithOpConversion final : public OpConversionPattern<ArithOp> {
...
+
+class SelectOpConversion : public OpConversionPattern<arith::SelectOp> {
+public:
+  using OpConversionPattern<arith::SelectOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(arith::SelectOp selectOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    auto dstType = getTypeConverter()->convertType(selectOp.getType());
```
Comment (simon-camp) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:66]:
```suggestion
    Type dstType = getTypeConverter()->convertType(selectOp.getType());
```

**Review by simon-camp — APPROVED** [review]
I'd replace the `auto` with `Type`, otherwise looks good.

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 3]
```diff
@@ -70,7 +95,8 @@ void mlir::populateArithToEmitCPatterns(TypeConverter &typeConverter,
     ArithOpConversion<arith::AddFOp, emitc::AddOp>,
     ArithOpConversion<arith::DivFOp, emitc::DivOp>,
     ArithOpConversion<arith::MulFOp, emitc::MulOp>,
-    ArithOpConversion<arith::SubFOp, emitc::SubOp>
+    ArithOpConversion<arith::SubFOp, emitc::SubOp>,
+    SelectOpConversion
```
Comment (TinaAMD) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:99]:
```suggestion
    ArithSelectOpConversion
```
Maybe rename it such that is more consistent to the rest?

**Review by TinaAMD — APPROVED** [review]

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp (hunk 3 above)
Comment (marbre) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:99]: The intend behind the naming `ArithOp` is that this is a conversion for a "genric" arith op. The `ArithToEmitC` already indicates that this file only contains conversions from the Arith to the EmitC dialect.

**Pull request merged**